pkg/controller/controller_instance.go

```diff
@@ -1457,35 +1466,52 @@ func (c *controller) processProvisionSuccess(instance *v1beta1.ServiceInstance,
 	return nil
 }
 
-// processProvisionFailure handles the logging and updating of a
+// processTerminalProvisionFailure handles the logging and updating of a
 // ServiceInstance that hit a terminal failure during provision reconciliation.
-func (c *controller) processProvisionFailure(instance *v1beta1.ServiceInstance, readyCond, failedCond *v1beta1.ServiceInstanceCondition, shouldMitigateOrphan bool) error {
-	currentReconciledGeneration := instance.Status.ReconciledGeneration
+func (c *controller) processTerminalProvisionFailure(instance *v1beta1.ServiceInstance, readyCond, failedCond *v1beta1.ServiceInstanceCondition, shouldMitigateOrphan bool) error {
 	if failedCond == nil {
 		return fmt.Errorf("failedCond must not be nil")
 	}
+	return c.processProvisionFailure(instance, readyCond, failedCond, shouldMitigateOrphan)
+}
+
+// processTemporaryProvisionFailure handles the logging and updating of a
+// ServiceInstance that hit a temporary error during provision reconciliation.
+func (c *controller) processTemporaryProvisionFailure(instance *v1beta1.ServiceInstance, readyCond *v1beta1.ServiceInstanceCondition, shouldMitigateOrphan bool) error {
+	return c.processProvisionFailure(instance, readyCond, nil, shouldMitigateOrphan)
+}
+
+// processProvisionFailure handles the logging and updating of a
+// ServiceInstance that hit a temporary or a terminal failure during provision
+// reconciliation.
+func (c *controller) processProvisionFailure(instance *v1beta1.ServiceInstance, readyCond, failedCond *v1beta1.ServiceInstanceCondition, shouldMitigateOrphan bool) error {
+	currentReconciledGeneration := instance.Status.ReconciledGeneration
 
 	if readyCond != nil {
 		c.recorder.Event(instance, corev1.EventTypeWarning, readyCond.Reason, readyCond.Message)
 		setServiceInstanceCondition(instance, v1beta1.ServiceInstanceConditionReady, readyCond.Status, readyCond.Reason, readyCond.Message)
 	}
 
-	c.recorder.Event(instance, corev1.EventTypeWarning, failedCond.Reason, failedCond.Message)
-	setServiceInstanceCondition(instance, v1beta1.ServiceInstanceConditionFailed, failedCond.Status, failedCond.Reason, failedCond.Message)
+	var errorMessage error
+	if failedCond != nil {
+		c.recorder.Event(instance, corev1.EventTypeWarning, failedCond.Reason, failedCond.Message)
+		setServiceInstanceCondition(instance, v1beta1.ServiceInstanceConditionFailed, failedCond.Status, failedCond.Reason, failedCond.Message)
+		errorMessage = fmt.Errorf(failedCond.Message)
+	} else {
+		resetServiceInstanceCondition(instance, v1beta1.ServiceInstanceConditionFailed)
```

Comment:
Why is resetting the Failed condition part of processing a failure? The Failed condition should be cleared out as part of starting a new provision rather than as part of finishing a provision. As it is, the Failed condition is not cleared out when the provision is successful.

Reply:
The `failedCond == nil` here means that we have encountered a retriable error (see the `processTemporaryProvisionFailure` method).
The `Failed:True` condition along with the latest `ObservedGeneration` means: "terminal failure that doesn't need retrying until the spec is updated (and consequently the Generation is incremented)".

As a reminder from #1748:
```go
+// isServiceInstanceProcessedAlready returns true if there is no further processing
+// needed for the instance based on ObservedGeneration
+func isServiceInstanceProcessedAlready(instance *v1beta1.ServiceInstance) bool {
+	return instance.Status.ObservedGeneration >= instance.Generation &&
+		(isServiceInstanceReady(instance) || isServiceInstanceFailed(instance)) && !instance.Status.OrphanMitigationInProgress
+}
```
The `isServiceInstanceProcessedAlready` is invoked before we start a new provisioning operation, and it will return `true` unless we clear the reset the `Failed` condition as I do there.

I don't like that this also means that we lose the `reason` and `message` there, but currently we store the same reason and message in the `Ready` condition, so it's probably not too bad.